src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,207 @@
+                          HttpConnectionEvents callbacks) {
+        if (uri == null) {  throw new IllegalArgumentException("URI must not be null"); }
+        if (uri.getScheme() == null) { throw new IllegalArgumentException("URI does not have a Scheme"); }
+        if (!HTTP.equals(uri.getScheme()) && !HTTPS.equals(uri.getScheme())) { throw new IllegalArgumentException("URI has unknown Scheme"); }
+        if (uri.getHost() == null) { throw new IllegalArgumentException("URI does not have a Host name"); }
+        if (bootstrap == null || bootstrap.isNull()) {  throw new IllegalArgumentException("ClientBootstrap must not be null"); }
+        if (socketOptions == null || socketOptions.isNull()) { throw new IllegalArgumentException("SocketOptions must not be null"); }
+        if (HTTPS.equals(uri.getScheme()) && tlsContext == null) { throw new IllegalArgumentException("TlsContext must not be null if https is used"); }
+
+        int port = uri.getPort();
+
+        /* Pick a default port based on the scheme if one wasn't set in the URI */
+        if (port == -1) {
+            if (HTTP.equals(uri.getScheme()))  { port = DEFAULT_HTTP_PORT; }
+            if (HTTPS.equals(uri.getScheme())) { port = DEFAULT_HTTPS_PORT; }
+        }
+
+        if (port <= 0 || 65535 < port) {
+            throw new IllegalArgumentException("Invalid or missing port: " + uri);
+        }
+
+        this.uri = uri;
+        this.port = port;
+        this.useTls = HTTPS.equals(uri.getScheme());
+        this.clientBootstrap = bootstrap;
+        this.socketOptions = socketOptions;
+        this.tlsContext = tlsContext;
+        this.userConnectionCallbacks = callbacks;
+        this.connectedFuture = new CompletableFuture<>();
+        this.shutdownFuture = new CompletableFuture<>();
+        this.connectAck = AsyncCallback.wrapFuture(connectedFuture, false);
+    }
+
+    /**
+     * Connects to the Http Endpoint
+     * @return Future indicating when
+     */
+    public CompletableFuture<Boolean> connect() throws CrtRuntimeException {
+        if (!isNull()) {
+            return connectedFuture;
+        }
+
+        acquire(httpConnectionNew(this,
+                clientBootstrap.native_ptr(),
+                socketOptions.native_ptr(),
+                useTls ? tlsContext.native_ptr() : 0,
+                uri.getHost(),
+                port));
+
+        return connectedFuture;
+    }
+
+    /**
+     * Disconnects if necessary, and frees native resources associated with this connection
+     */
+    @Override
+    public void close() {
+        if (!isNull()) {
+            httpConnectionRelease(release());
+        }
+
+        super.close();
+    }
+
+    /** Called from native when the connection is established the first time **/
+    private void onConnectionComplete(int errorCode) {
+        if (userConnectionCallbacks != null) {
+            userConnectionCallbacks.onConnectionComplete(errorCode);
```

Comment:
There should not be user connection callbacks in addition to the future, the future is enough. Otherwise, you put the user in a position where they get the same information twice.

Reply:
you can put a callback on the future anyways. Agreed, don't mix futures and callbacks.

Reply:
Fixed